File: k8s/resolver.py
# Copyright 2017 The Bazel Authors. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Walks a yaml object and resolves all docker_name.Tag to docker_name.Digest.
"""

from __future__ import print_function

import argparse
import logging
import os
import sys

from containerregistry.client import docker_creds
from containerregistry.client import docker_name
from containerregistry.client.v2 import docker_image as v2_image
from containerregistry.client.v2_2 import docker_image as v2_2_image
from containerregistry.client.v2_2 import docker_session as v2_2_session
from containerregistry.tools import patched
from containerregistry.transport import transport_pool

import httplib2
import yaml


parser = argparse.ArgumentParser(
    description='Resolve image references to digests.')

parser.add_argument(
  '--template', action='store',
  help='The template file to resolve.')

parser.add_argument(
  '--image_spec', action='append',
  help='Associative lists of the constitutent elements of a FromDisk image.')

parser.add_argument(
  '--image_chroot', action='store',
  help=('The repository under which to chroot image references when '
        'publishing them.'))

_THREADS = 32


def Resolve(input, string_to_digest):
  """Translate tag references within the input yaml into digests."""
  def walk_dict(d):
    return {
      walk(k): walk(v)
      for (k, v) in d.iteritems()
    }

  def walk_list(l):
    return [walk(e) for e in l]

  def walk_string(s):
    try:
      return string_to_digest(s)
    except:
      return s

  def walk(o):
    if isinstance(o, dict):
      return walk_dict(o)
    if isinstance(o, list):
      return walk_list(o)
    if isinstance(o, str):
      return walk_string(o)
    return o

  return yaml.dump_all(map(walk, yaml.load_all(input)))


def StringToDigest(string, overrides, transport):
  """Turn a string into a stringified digest."""
  if string in overrides:
    return str(overrides[string])

  # Attempt to turn the string into a tag, this may throw.
  tag = docker_name.Tag(string)

  def fully_qualify_digest(digest):
    return docker_name.Digest('{registry}/{repo}@{digest}'.format(
      registry=tag.registry, repo=tag.repository, digest=digest))

  # Resolve the tag to digest using the standard
  # Docker keychain logic.
  creds = docker_creds.DefaultKeychain.Resolve(tag)
  with v2_2_image.FromRegistry(tag, creds, transport) as img:
    if img.exists():
      digest = fully_qualify_digest(img.digest())
      overrides[string] = digest
      return str(digest)

  # If the tag doesn't exists as v2.2, then try as v2.
  with v2_image.FromRegistry(tag, creds, transport) as img:
    digest = fully_qualify_digest(img.digest())
    overrides[string] = digest
    return str(digest)


def Publish(transport, image_chroot,
            name=None, tarball=None, config=None, digest=None, layer=None):
  if not name:
    raise Exception('Expected "name" kwarg')

  if not config and (layer or digest):
    raise Exception(
      name + ': Using "layer" or "digest" requires "config" to be specified.')

  if config:
    with open(config, 'r') as reader:
      config = reader.read()
  elif tarball:
    with v2_2_image.FromTarball(tarball) as base:
      config = base.config_file()
  else:
    raise Exception(name + ': Either "config" or "tarball" must be specified.')

  if digest or layer:
    digest = digest.split(',')
    layer = layer.split(',')
    if len(digest) != len(layer):
      raise Exception(
          name + ': "digest" and "layer" must have matching lengths.')
  else:
    digest = []
    layer = []

  name_to_replace = name
  if image_chroot:
    name_to_publish = docker_name.Tag(os.path.join(image_chroot, name), strict=False)
  else:
    # Without a chroot, the left-hand-side must be a valid tag.
    name_to_publish = docker_name.Tag(name, strict=False)

  # Resolve the appropriate credential to use based on the standard Docker
  # client logic.
  creds = docker_creds.DefaultKeychain.Resolve(name_to_publish)

  with v2_2_session.Push(name_to_publish, creds, transport, threads=_THREADS) as session:
    with v2_2_image.FromDisk(config, zip(digest or [], layer or []),
                             legacy_base=tarball) as v2_2_img:
      session.upload(v2_2_img)

      return (name_to_replace, docker_name.Digest('{repository}@{digest}'.format(
          repository=name_to_publish.as_repository(),
          digest=v2_2_img.digest())))


def main():
  args = parser.parse_args()

  transport = transport_pool.Http(httplib2.Http, size=_THREADS)

  unseen_strings = set()
  overrides = {}
  # TODO(mattmoor): Execute these in a threadpool and
  # aggregate the results as they complete.
  for spec in args.image_spec or []:
    parts = spec.split(';')
    kwargs = dict([x.split('=', 2) for x in parts])
    try:
      (tag, digest) = Publish(transport, args.image_chroot, **kwargs)
      overrides[tag] = digest
      unseen_strings.add(tag)
    except Exception as e:
      logging.fatal('Error publishing provided image: %s', e)

  with open(args.template, 'r') as f:
    inputs = f.read()

  def _StringToDigest(t):
    if t in unseen_strings:
      unseen_strings.remove(t)
    return StringToDigest(t, overrides, transport)

  content = Resolve(inputs, _StringToDigest)

  if len(unseen_strings) > 0:
    print('ERROR: The following image references were not found in %r:' %
          args.template, file=sys.stderr)
    for ref in unseen_strings:
      print('    %s' % ref, file=sys.stderr)
    sys.exit(1)

  print(content)


if __name__ == '__main__':
  with patched.Httplib2():
    main()

File: k8s/go/cmd/resolver/resolver.go
// Copyright 2017 The Bazel Authors. All rights reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
package main

import (
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"path"
	"reflect"
	"strings"

	"gopkg.in/yaml.v2"

	"github.com/bazelbuild/rules_docker/container/go/pkg/compat"
	"github.com/bazelbuild/rules_docker/container/go/pkg/utils"
	"github.com/google/go-containerregistry/pkg/authn"
	"github.com/google/go-containerregistry/pkg/name"
	"github.com/google/go-containerregistry/pkg/v1/remote"
)

var (
	imgChroot         = flag.String("image_chroot", "", "The repository under which to chroot image references when publishing them.")
	k8sTemplate       = flag.String("template", "", "The k8s YAML template file to resolve.")
	allowUnusedImages = flag.Bool("allow_unused_images", false, "Allow images that don't appear in the JSON. This is useful when generating multiple SKUs of a k8s_object, only some of which use a particular image.")
	stampInfoFile     utils.ArrayStringFlags
	imgSpecs          utils.ArrayStringFlags
)

// imageSpec describes the differents parts of an image generated by
// rules_docker.
type imageSpec struct {
	// name is the name of the image.
	name string
	// imgTarball is the image in the `docker save` tarball format.
	imgTarball string
	// imgConfig if the config JSON file of the image.
	imgConfig string
	// digests is a list of files with the sha256 digests of the compressed
	// layers.
	digests []string
	// diffIDs is a list of files with the sha256 digests of the uncompressed
	// layers.
	diffIDs []string
	// compressedLayers are the paths to the compressed layer tarballs.
	compressedLayers []string
	// uncompressedLayers are the paths to the uncompressed layer tarballs.
	uncomressedLayers []string
}

// layers returns a list of strings that can be passed to the image reader in
// the compatiblity package of rules_docker to read the layers of an image in
// the format "va11,val2,val3,val4" where:
// val1 is the compressed layer tarball.
// val2 is the uncompressed layer tarball.
// val3 is the digest file.
// val4 is the diffID file.
func (s *imageSpec) layers() ([]string, error) {
	result := []string{}
	if len(s.digests) != len(s.diffIDs) || len(s.diffIDs) != len(s.compressedLayers) || len(s.compressedLayers) != len(s.uncomressedLayers) {
		return nil, fmt.Errorf("digest, diffID, compressed blobs & uncompressed blobs had unequal lengths for image %s, got %d, %d, %d, %d, want all of the lengths to be equal", s.name, len(s.digests), len(s.diffIDs), len(s.compressedLayers), len(s.uncomressedLayers))
	}
	for i, digest := range s.digests {
		diffID := s.diffIDs[i]
		compressedLayer := s.compressedLayers[i]
		uncompressedLayer := s.uncomressedLayers[i]
		result = append(result, fmt.Sprintf("%s,%s,%s,%s", compressedLayer, uncompressedLayer, digest, diffID))
	}
	return result, nil
}

// parseImageSpec parses the differents parts of a single docker image specified
// as string in the format "key1=val1;key2=val2" where the expected keys are:
// 1. "name": Name of the image.
// 2. "tarball": docker save tarball of the image.
// 3. "config": JSON config file of the image.
// 4. "diff_id": Files with sha256 digest of uncompressed layers.
// 5. "digest": Files with sha256 digest of compressed layers.
// 6. "compressed_layer": Path to compressed layer tarballs.
// 7. "uncompressed_layer": Path to uncompressed layer tarballs.
func parseImageSpec(spec string) (imageSpec, error) {
	result := imageSpec{}
	splitSpec := strings.Split(spec, ";")
	for _, s := range splitSpec {
		splitFields := strings.SplitN(s, "=", 2)
		if len(splitFields) != 2 {
			return imageSpec{}, fmt.Errorf("image spec item %q split by '=' into unexpected fields, got %d, want 2", s, len(splitFields))
		}
		switch splitFields[0] {
		case "name":
			result.name = splitFields[1]
		case "tarball":
			result.imgTarball = splitFields[1]
		case "config":
			result.imgConfig = splitFields[1]
		case "diff_id":
			result.diffIDs = strings.Split(splitFields[1], ",")
		case "digest":
			result.digests = strings.Split(splitFields[1], ",")
		case "compressed_layer":
			result.compressedLayers = strings.Split(splitFields[1], ",")
		case "uncompressed_layer":
			result.uncomressedLayers = strings.Split(splitFields[1], ",")
		default:
			return imageSpec{}, fmt.Errorf("unknown image spec field %q", splitFields[0])
		}
	}
	return result, nil
}

// publishSingle publishes a docker image with the given spec to the remote
// registry indicated in the image name. The image name is stamped with the
// given stamper.
// The stamped image name is returned referenced by its sha256 digest.
func publishSingle(spec imageSpec, stamper *compat.Stamper) (string, error) {
	layers, err := spec.layers()
	if err != nil {
		return "", fmt.Errorf("unable to convert the layer parts in image spec for %s into a single comma separated argument: %v", spec.name, err)
	}

	imgParts, err := compat.ImagePartsFromArgs(spec.imgConfig, "", spec.imgTarball, layers)
	if err != nil {
		return "", fmt.Errorf("unable to determine parts of the image from the specified arguments: %v", err)
	}
	img, err := compat.ReadImage(imgParts)
	if err != nil {
		return "", fmt.Errorf("error reading image: %v", err)
	}
	stampedName := stamper.Stamp(spec.name)

	var ref name.Reference
	if *imgChroot != "" {
		n := path.Join(*imgChroot, stampedName)
		t, err := name.NewTag(n, name.WeakValidation)
		if err != nil {
			return "", fmt.Errorf("unable to create a docker tag from stamped name %q: %v", n, err)
		}
		ref = t
	} else {
		t, err := name.NewTag(stampedName, name.WeakValidation)
		if err != nil {
			return "", fmt.Errorf("unable to create a docker tag from stamped name %q: %v", stampedName, err)
		}
		ref = t
	}
	auth, err := authn.DefaultKeychain.Resolve(ref.Context())
	if err != nil {
		return "", fmt.Errorf("unable to get authenticator for image %v", ref.Name())
	}

	if err := remote.Write(ref, img, remote.WithAuth(auth)); err != nil {
		return "", fmt.Errorf("unable to push image %v: %v", ref.Name(), err)
	}

	d, err := img.Digest()
	if err != nil {
		return "", fmt.Errorf("unable to get digest of image %v", ref.Name())
	}

	return fmt.Sprintf("%s/%s@%v", ref.Context().RegistryStr(), ref.Context().RepositoryStr(), d), nil
}

// publish publishes the image with the given spec. It returns:
// 1. A map from the unstamped & tagged image name to the stamped image name
//    referenced by its sha256 digest.
// 2. A set of unstamped & tagged image names that were pushed to the registry.
func publish(spec []imageSpec, stamper *compat.Stamper) (map[string]string, map[string]bool, error) {
	overrides := make(map[string]string)
	unseen := make(map[string]bool)
	for _, s := range spec {
		digestRef, err := publishSingle(s, stamper)
		if err != nil {
			return nil, nil, fmt.Errorf("unable to publish image %s", s.name)
		}
		overrides[s.name] = digestRef
		unseen[s.name] = true
	}
	return overrides, unseen, nil
}

type resolver struct {
	resolvedImages map[string]string
	unseen         map[string]bool
}

func (r *resolver) resolveItem(i interface{}) (interface{}, error) {
	if s, ok := i.(string); ok {
		log.Printf("Resolve item: %q", s)
		return s, nil
	}
	if l, ok := i.([]interface{}); ok {
		return r.resolveList(l)
	}
	if m, ok := i.(map[interface{}]interface{}); ok {
		return r.resolveMap(m)
	}
	log.Printf("Fallthrough resolving %v of type %v.", i, reflect.TypeOf(i))
	return i, nil
}

func (r *resolver) resolveList(l []interface{}) ([]interface{}, error) {
	result := []interface{}{}
	for _, i := range l {
		o, err := r.resolveItem(i)
		if err != nil {
			return nil, err
		}
		result = append(result, o)
	}
	return result, nil
}

func (r *resolver) resolveMap(m map[interface{}]interface{}) (map[interface{}]interface{}, error) {
	result := make(map[interface{}]interface{})
	for k, v := range m {
		rk, err := r.resolveItem(k)
		if err != nil {
			return nil, err
		}
		rv, err := r.resolveItem(v)
		if err != nil {
			return nil, err
		}
		result[rk] = rv
	}
	return result, nil
}

func (r *resolver) walkYAML(b []byte) error {
	var l []interface{}
	lErr := yaml.Unmarshal(b, &l)
	if lErr == nil {
		r.resolveItem(l)
		return nil
	}
	var m map[interface{}]interface{}
	mErr := yaml.Unmarshal(b, &m)
	if mErr == nil {
		r.resolveMap(m)
		return nil
	}

	return fmt.Errorf("unable to parse given blob as a YAML list or map: %v %v", lErr, mErr)
}

func resolveTemplate(templateFile string, resolvedImages map[string]string, unseen map[string]bool) error {
	t, err := ioutil.ReadFile(templateFile)
	if err != nil {
		return fmt.Errorf("unable to open template file %q: %v", templateFile, err)
	}

	r := resolver{
		resolvedImages: resolvedImages,
		unseen:         unseen,
	}
	if r.walkYAML(t); err != nil {
		return fmt.Errorf("unable to resolve YAML template %q: %v", templateFile, err)
	}
	return nil
}

func main() {
	flag.Var(&imgSpecs, "image_spec", "Associative lists of the constitutent elements of a docker image.")
	flag.Var(&stampInfoFile, "stamp-info-file", "One or more Bazel stamp info files.")
	flag.Parse()

	log.Println("Template", *k8sTemplate)

	stamper, err := compat.NewStamper(stampInfoFile)
	if err != nil {
		log.Fatalf("Failed to initialize the stamper: %v", err)
	}

	specs := []imageSpec{}
	for _, s := range imgSpecs {
		spec, err := parseImageSpec(s)
		if err != nil {
			log.Fatalf("Unable to parse image spec %q: %v", s, err)
		}
		specs = append(specs, spec)
	}
	resolvedImages, unseen, err := publish(specs, stamper)
	if err != nil {
		log.Fatalf("Unable to publish images: %v", err)
	}
	if err := resolveTemplate(*k8sTemplate, resolvedImages, unseen); err != nil {
		log.Fatalf("Unable to resolve template file %q: %v", *k8sTemplate, err)
	}
	if len(unseen) > 0 && !*allowUnusedImages {
		log.Printf("The following images given as --image_spec were not found in the template:")
		for i := range unseen {
			log.Printf("%s", i)
		}
		log.Fatalf("--allow_unused_images can be specified to ignore this error.")
	}
}
